TechnicalDocuments/Proposals/TrackScreenEventsAsSideEffects.md

```diff
@@ -0,0 +1,197 @@
+possible to define a generic mechanism for posting screen views to the analytics
+service.
+
+```swift
+enum FlowAnalyticsEvent: AnalyticsEvent {
+    case screenView(name: String)
+}
+
+extension FlowAnalyticsEvent {
+    static func screenViewTracking(
+        _ flow: Flow,
+        with tracker: AnalyticsTrackingService
+    ) -> Flow {
+
+        func trackScreenView() {
+            guard let viewController = flow.currentViewController() else { return }
+
+            if let screenName = viewController.screenName(),
+                screenName.isNotEmpty {
+                tracker.track(FlowAnalyticsEvent.screenView(name: screenName))
+            }
+        }
+
+        func decoratedPresent(_ viewController: UIViewController, _ animated: Bool, _ completion: (() -> Void)?) {
+            flow.present(viewController, animated: animated, completion: completion)
+            trackScreenView()
+        }
+
+        func decoratedReplace(_ viewController: UIViewController, _ animated: Bool) {
+            flow.replace(with: viewController, animated: animated)
+            trackScreenView()
+        }
+
+        func decoratedDismiss(animated: Bool, _ completion: (() -> Void)?) {
+            flow.dismiss(animated: animated, completion: completion)
+            trackScreenView()
+        }
+
+        return Flow(
+            presentViewController: decoratedPresent,
+            replaceViewController: decoratedReplace,
+            dismissViewController: decoratedDismiss,
+            currentViewController: flow.currentViewController)
+    }
+
+    static func screenViewTracking(_ flow: Flow) -> Flow {
+        return screenViewTracking(flow, with: Current.analyticsService)
+    }
+}
+```
+
+This example uses the `Flow` implementation that has been refactored to a struct
+that I have suggested in a previous proposal. Something similar can be done with
+the current `protocol` implementation, but to me it feels more natural to use a
+struct with function variables. Please note that `currentViewController()` is
+needed to post a screen event when a modal is dismissed or the user goes back
+in a navigation stack.
+
+Finally, to track screen view events we need to decorate the `Flow`, here is
+how to do it for the GP @ Hand Introduction journey
+
+```swift
+private func showNHSIntro() {
+    BabylonNavigationController { navigation, modal in
+        return builders.nhsOnboarding.make(
+            navigation: navigation |> FlowAnalyticsEvent.screenViewTracking,
+            modal: modal |> FlowAnalyticsEvent.screenViewTracking,
+            parent: self.modal,
+            root: self.modal
+        )
+    } |> modal.present
+}
+```
+
+I added this helper function to make `|>` available.
+
+```swift
+static func screenViewTracking(_ flow: Flow) -> Flow {
+    return screenViewTracking(flow, with: Current.analyticsService)
+}
+```
+
+It is important to note that `Flow` instances should only be decorated at the
```

Comment:
This can be error prone, if such mistake is done then it will result in double tracking. Can we maybe make it a bit safer by doing something like this:

```swift
extension UIViewController {
    struct Tracking {
       let vc: UIViewController
       var modalFlow: Flow {
          return FlowAnalyticsEvent.screenViewTracking(vc.modalFlow)
       }
    }
    var tracking: Tracking {
       return Tracking(vc: self)
    }
}
```

Then it will not be possible to decorate flow anywhere outside `BabylonUI`, for that `screenViewTracking` should be made internal. The usage will be then

```
vc.tracking.modalFlow -> flow with tracking
vc.modalFlow -> flow without tracking (if we ever need to suppress tracking in any parts of the app)
```

What do you think?

Reply:
I don't mind experimenting with designs that make it less likely that we get bugs from double posting. 

At the moment I am working on creating a mock analytics service that we can use in our UI tests and unit tests. I am hoping that this will make it inexpensive to include analytics in our automation tests.